snitun/server/run.py

```diff
@@ -303,58 +308,59 @@ def _register_parial(fileno: int, data: bytes) -> bool:
 
     def _process(
         self,
-        con: socket.socket,
+        client: Connection,
         workers_lb: Iterable[ServerWorker],
-        data: bytes,
-    ) -> None | bytes:
+    ) -> None:
         """Process connection & helo."""
         try:
-            data += con.recv(MAX_READ_SIZE)
+            client.buffer += client.sock.recv(MAX_READ_SIZE)
         except OSError as err:
             _LOGGER.warning("Receive fails: %s", err)
-            self._close_socket(con, shutdown=False)
-            return None
+            client.close_socket(shutdown=False)
+            return
 
         # No data received
-        if not data:
-            self._close_socket(con)
-            return None
+        if not client.buffer:
+            client.close_socket()
+            return
 
         # Peer connection
-        if data.startswith(b"gA"):
-            next(workers_lb).handover_connection(con, data)
-            _LOGGER.debug("Handover new peer connection: %s", data)
-            return None
+        if client.buffer.startswith(b"gA"):
+            next(workers_lb).handover_connection(client.sock, client.buffer)
+            _LOGGER.debug("Handover new peer connection: %s", client.buffer)
+            client.soft_close()
+            return
 
         # TLS/SSL connection
-        if data[0] != 0x16:
-            _LOGGER.warning("No valid ClientHello found: %s", data)
-            self._close_socket(con)
-            return None
+        if client.buffer[0] != 0x16:
+            _LOGGER.warning("No valid ClientHello found: %s", client.buffer)
+            client.close_socket()
+            return
 
+        # Get Hostname
         try:
-            hostname = parse_tls_sni(data)
+            hostname = parse_tls_sni(client.buffer)
         except ParseSNIIncompleteError:
-            return data
+            # Check Buffer Size
+            if len(client.buffer) >= MAX_BUFFER_SIZE:
+                _LOGGER.warning("Connection %d exceed buffer size", client.fileno)
+                client.close_socket()
+            return
         except ParseSNIError:
             _LOGGER.warning("Receive invalid ClientHello on public Interface")
-            return None
+            client.close_socket()
+            return
+
+        # Distribute to child worker
         for worker in self._workers:
             if not worker.is_responsible_peer(hostname):
                 continue
-            worker.handover_connection(con, data, sni=hostname)
+            worker.handover_connection(client.sock, client.buffer, sni=hostname)
+            client.soft_close()
```

Comment:
Shouldn't this happen before the handover?
Like before?

Reply:
Doesn't matters as we take it out of the epoll before the next circle. But I can change it